# Prática 1: Limpieza de datos
#### Aplicar funciones para resumir y entender los datos (Summary).
#### Componer datos, eliminar problemas en datos NA, NULL, revisar la primer hoja del Road-Accident-Safety-Data-Guide-1979-2004.xls para conocer las consideraciones (Fix)
#### Cambiar los tipos de datos necesarios (fechas, números con los que se vayan a operar, etc.), crear los factores de acuerdo a Road-Accident-Safety-Data-Guide-1979-2004.xls (Fix)
#### Si es necesario, adaptar los atributos que sean necesarios (Adapt).

## accidents2004
### Funciones para resumir y entender los datos:

In [3]:
# importamos pandas para el análisis y manipulación de los datos
import pandas as pd

In [4]:
# Leemos el csv
accidents = pd.read_csv('../data/accidents2004.csv')

In [6]:
accidents.info() # nos devuelve la información del DateFrame, como el número de filas y columnas,
# los nombres de las columnas así como los tipos de cada una (tipo de dato) y la cantidad memoria usada
# por el DataFrame

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207410 entries, 0 to 207409
Data columns (total 33 columns):
 #   Column                                       Non-Null Count   Dtype  
---  ------                                       --------------   -----  
 0   Unnamed: 0                                   207410 non-null  int64  
 1   �..Accident_Index                            207410 non-null  object 
 2   Location_Easting_OSGR                        207280 non-null  float64
 3   Location_Northing_OSGR                       207280 non-null  float64
 4   Longitude                                    207280 non-null  float64
 5   Latitude                                     207280 non-null  float64
 6   Police_Force                                 207410 non-null  int64  
 7   Accident_Severity                            207410 non-null  int64  
 8   Number_of_Vehicles                           207410 non-null  int64  
 9   Number_of_Casualties                         207410 non-nul

In [7]:
# mostramos los primeros registros con el método head
accidents.head()

Unnamed: 0 �..Accident_Index  Location_Easting_OSGR  \
0     6016789     200401BS00001               526880.0   
1     6016790     200401BS00002               526230.0   
2     6016791     200401BS00003               526400.0   
3     6016792     200401BS00004               524260.0   
4     6016793     200401BS00006               523700.0   

   Location_Northing_OSGR  Longitude   Latitude  Police_Force  \
0                179000.0  -0.173622  51.495658             1   
1                178450.0  -0.183177  51.490860             1   
2                179620.0  -0.180311  51.501337             1   
3                179470.0  -0.211179  51.500464             1   
4                181470.0  -0.218542  51.518561             1   

   Accident_Severity  Number_of_Vehicles  Number_of_Casualties  ...  \
0                  3                   2                     1  ...   
1                  2                   1                     1  ...   
2                  3                   2                     2  ...   
3                  3                   2                     1  ...   
4                  2                   2                     1  ...   

  Pedestrian_Crossing.Human_Control  Pedestrian_Crossing.Physical_Facilities  \
0                                 0                                        5   
1                                 0                                        5   
2                                 0                                        5   
3                                 0                                        0   
4                                 0                                        0   

  Light_Conditions  Weather_Conditions Road_Surface_Conditions  \
0                4                   1                       1   
1                4                   2                       2   
2                4                   2                       2   
3                1                   2                       2   
4                1                   1                       1   

   Special_Conditions_at_Site  Carriageway_Hazards  Urban_or_Rural_Area  \
0                           0                    0                    1   
1                           0                    0                    1   
2                           0                    0                    1   
3                           0                    0                    1   
4                           0                    0                    1   

   Did_Police_Officer_Attend_Scene_of_Accident  LSOA_of_Accident_Location  
0                                            1                  E01002821  
1                                            1                  E01002836  
2                                            1                  E01002891  
3                                            1                  E01002866  
4                                            1                  E01002876  

[5 rows x 33 columns]

In [8]:
# y los últimos registros con el método tail
accidents.tail()

Unnamed: 0 �..Accident_Index  Location_Easting_OSGR  \
207405     6224194     2004984162204               322340.0   
207406     6224195     2004984162804               317850.0   
207407     6224196     2004984163904               304300.0   
207408     6224197     2004984164304               306400.0   
207409     6224198     2004984164804               320150.0   

        Location_Northing_OSGR  Longitude   Latitude  Police_Force  \
207405                567270.0  -3.215400  54.993955            98   
207406                567570.0  -3.285649  54.995929            98   
207407                607980.0  -3.511052  55.356510            98   
207408                605130.0  -3.476987  55.331314            98   
207409                569440.0  -3.250229  55.013104            98   

        Accident_Severity  Number_of_Vehicles  Number_of_Casualties  ...  \
207405                  3                   3                     1  ...   
207406                  3                   1                     1  ...   
207407                  3                   1                     1  ...   
207408                  2                   1                     2  ...   
207409                  3                   1                     1  ...   

       Pedestrian_Crossing.Human_Control  \
207405                                 0   
207406                                 0   
207407                                 0   
207408                                 0   
207409                                 0   

        Pedestrian_Crossing.Physical_Facilities Light_Conditions  \
207405                                        0                1   
207406                                        0                4   
207407                                        0                6   
207408                                        0                6   
207409                                        0                6   

        Weather_Conditions Road_Surface_Conditions  \
207405                   2                       4   
207406                   4                       2   
207407                   8                       4   
207408                   5                       2   
207409                   1                       2   

        Special_Conditions_at_Site  Carriageway_Hazards  Urban_or_Rural_Area  \
207405                           0                    0                    2   
207406                           0                    0                    2   
207407                           0                    0                    2   
207408                           0                    0                    2   
207409                           0                    0                    2   

        Did_Police_Officer_Attend_Scene_of_Accident  LSOA_of_Accident_Location  
207405                                            1                        NaN  
207406                                            1                        NaN  
207407                                            1                        NaN  
207408                                            1                        NaN  
207409                                            1                        NaN  

[5 rows x 33 columns]

In [9]:
# Obtenemos una descripción general de estadísticos con el método describe
accidents.describe()

Unnamed: 0  Location_Easting_OSGR  Location_Northing_OSGR  \
count  2.074100e+05          207280.000000            2.072800e+05   
mean   6.120494e+06          436584.834765            3.022250e+05   
std    5.987425e+04           95293.013359            1.603311e+05   
min    6.016789e+06           65420.000000            1.027000e+04   
25%    6.068641e+06          372000.000000            1.792900e+05   
50%    6.120494e+06          435900.000000            2.770100e+05   
75%    6.172346e+06          520540.000000            3.986700e+05   
max    6.224198e+06          655230.000000            1.201800e+06   

           Longitude       Latitude   Police_Force  Accident_Severity  \
count  207280.000000  207280.000000  207410.000000      207410.000000   
mean       -1.480996      52.608144      30.710038           2.842322   
std         1.401643       1.443622      25.572253           0.401911   
min        -7.506902      49.912761       1.000000           1.000000   
25%        -2.420322      51.500292       7.000000           3.000000   
50%        -1.462320      52.384309      30.000000           3.000000   
75%        -0.252341      53.483442      46.000000           3.000000   
max         1.757968      60.695193      98.000000           3.000000   

       Number_of_Vehicles  Number_of_Casualties    Day_of_Week  ...  \
count       207410.000000         207410.000000  207410.000000  ...   
mean             1.831373              1.354033       4.137689  ...   
std              0.719767              0.830056       1.929810  ...   
min              1.000000              1.000000       1.000000  ...   
25%              1.000000              1.000000       3.000000  ...   
50%              2.000000              1.000000       4.000000  ...   
75%              2.000000              1.000000       6.000000  ...   
max             18.000000             67.000000       7.000000  ...   

       X2nd_Road_Number  Pedestrian_Crossing.Human_Control  \
count     207410.000000                      207410.000000   
mean         390.265060                           0.005178   
std         1303.867462                           0.092703   
min           -1.000000                          -1.000000   
25%            0.000000                           0.000000   
50%            0.000000                           0.000000   
75%            0.000000                           0.000000   
max         9999.000000                           2.000000   

       Pedestrian_Crossing.Physical_Facilities  Light_Conditions  \
count                            207410.000000     207410.000000   
mean                                  0.645065          1.985184   
std                                   1.711374          1.660734   
min                                  -1.000000          1.000000   
25%                                   0.000000          1.000000   
50%                                   0.000000          1.000000   
75%                                   0.000000          4.000000   
max                                   8.000000          7.000000   

       Weather_Conditions  Road_Surface_Conditions  \
count       207410.000000            207410.000000   
mean             1.498689                 1.365035   
std              1.463468                 0.618476   
min             -1.000000                -1.000000   
25%              1.000000                 1.000000   
50%              1.000000                 1.000000   
75%              1.000000                 2.000000   
max              9.000000                 5.000000   

       Special_Conditions_at_Site  Carriageway_Hazards  Urban_or_Rural_Area  \
count               207410.000000        207410.000000        207410.000000   
mean                     0.064322             0.016860             1.366646   
std                      0.504591             0.281127             0.483598   
min                     -1.000000            -1.000000             1.000000   
25%                      0

Con esto podemos ver que el data frame se encuentra conformado por 207410 filas por 33 columnas, las filas están
indexadas a partir del número 0, y las columnas son 4 de tipo float, 24 de tipo int, y las otras 5 corresponden al tipo object, además podemos observar que **no** se está haciendo del tipo de dato NaN para especificar datos nulos o faltantes, por lo que tendremos que recurrir limpiar los datos de forma no directa, podemos ver la cantidad de memoria que nos está ocupando cargar el .csv y por último mediante el método describe podemos ver algunos estadísticos importantes como el conteo, la media, desviación estándar, entre otro (recordar que esta información está sucia debido a los datos faltantes).

## Eliminar datos faltantes y nulos:

En la documentación para el fix se menciona que se hace uso de **-1** para los valores null o fuera del rango, por lo que nuestra tarea es limpiar estos datos

puesto que se está usando el valor -1 como indicador para los valores nulos o faltantes debemos buscar todas la filas
que contienen dicho valor:

In [25]:
missing_values = accidents[accidents.eq(-1).any(1)]
missing_values
# el método .eq(n) busca el número n, mientras que con .any(axis) le indicamos dónde buscar
# en Pandas axis=0 es usado para los renglones y axis=1 para las columnas, por lo que la línea
# de arriba busca el valor -1 en barriendo todas las columnas, es decir, si encuentra un -1 devuelve
# esa fila

Unnamed: 0 �..Accident_Index  Location_Easting_OSGR  \
0          6016789     200401BS00001               526880.0   
3          6016792     200401BS00004               524260.0   
9          6016798     200401BS00012               525970.0   
11         6016800     200401BS00014               524180.0   
12         6016801     200401BS00015               525970.0   
...            ...               ...                    ...   
207404     6224193     2004984161704               319400.0   
207405     6224194     2004984162204               322340.0   
207407     6224196     2004984163904               304300.0   
207408     6224197     2004984164304               306400.0   
207409     6224198     2004984164804               320150.0   

        Location_Northing_OSGR  Longitude   Latitude  Police_Force  \
0                     179000.0  -0.173622  51.495658             1   
3                     179470.0  -0.211179  51.500464             1   
9                     177930.0  -0.187106  51.486245             1   
11                    179570.0  -0.212296  51.501380             1   
12                    177770.0  -0.187163  51.484807             1   
...                        ...        ...        ...           ...   
207404                566580.0  -3.261148  54.987289            98   
207405                567270.0  -3.215400  54.993955            98   
207407                607980.0  -3.511052  55.356510            98   
207408                605130.0  -3.476987  55.331314            98   
207409                569440.0  -3.250229  55.013104            98   

        Accident_Severity  Number_of_Vehicles  Number_of_Casualties  ...  \
0                       3                   2                     1  ...   
3                       3                   2                     1  ...   
9                       3                   2                     1  ...   
11                      3                   2                     1  ...   
12                      3                   3                     2  ...   
...                   ...                 ...                   ...  ...   
207404                  3                   1                     1  ...   
207405                  3                   3                     1  ...   
207407                  3                   1                     1  ...   
207408                  2                   1                     2  ...   
207409                  3                   1                     1  ...   

       Pedestrian_Crossing.Human_Control  \
0                                      0   
3                                      0   
9                                      0   
11                                     0   
12                                     0   
...                                  ...   
207404                                 0   
207405                                 0   
207407                                 0   
207408                                 0   
207409                                 0   

        Pedestrian_Crossing.Physical_Facilities Light_Conditions  \
0                                             5                4   
3                                             0                1   
9                                             0                1   
11                                            0                1   
12                                            0                4   
...                                         ...              ...   
207404                                        0                1   
207405                                        0                1   
207407                                        0                6   
207408                                        0                6   
207409                                        0                6   

        Weather_Conditions Road_Surface_Conditions  \
0                        1                       1   
3                        2                 

podemos observar que tenemos un total 97500 filas que contienen al menos un valor -1, con esto ya encontramos los valores faltantes, en este caso elegiremos remover los datos faltantes puesto que la cantidad de datos es alta,
con esto recordemos que el data frame original cuenta con 207410 filas de las cuales 97500 cuentan con datos faltantes, por lo que nuestro data freme sin datos faltantes debe contar con 207410 - 97500 = **109910**

In [43]:
accidents = accidents[accidents.ne(-1).all(1)]
# en el ejemplo anterior buscamos cualquier fila que contuviera el valor -1, ahora buscamos todas las
# filas que no contienen el valor -1, usamos ne: not equal en lugar de eq y all en lugar de any

In [47]:
len(accidents) # podemos comprobar que el número de filas que obtenemos corresponde a la resta que calculamos

109910

Hasta este punto tenemos nuestro data frame libre de datos faltantes

In [49]:
accidents.head()

Unnamed: 0 �..Accident_Index  Location_Easting_OSGR  \
1     6016790     200401BS00002               526230.0   
2     6016791     200401BS00003               526400.0   
4     6016793     200401BS00006               523700.0   
5     6016794     200401BS00007               527400.0   
6     6016795     200401BS00009               524470.0   

   Location_Northing_OSGR  Longitude   Latitude  Police_Force  \
1                178450.0  -0.183177  51.490860             1   
2                179620.0  -0.180311  51.501337             1   
4                181470.0  -0.218542  51.518561             1   
5                178650.0  -0.166262  51.492396             1   
6                180140.0  -0.207919  51.506439             1   

   Accident_Severity  Number_of_Vehicles  Number_of_Casualties  ...  \
1                  2                   1                     1  ...   
2                  3                   2                     2  ...   
4                  2                   2                     1  ...   
5                  2                   2                     1  ...   
6                  3                   1                     1  ...   

  Pedestrian_Crossing.Human_Control  Pedestrian_Crossing.Physical_Facilities  \
1                                 0                                        5   
2                                 0                                        5   
4                                 0                                        0   
5                                 0                                        1   
6                                 0                                        0   

  Light_Conditions  Weather_Conditions Road_Surface_Conditions  \
1                4                   2                       2   
2                4                   2                       2   
4                1                   1                       1   
5                4                   1                       1   
6                4                   1                       1   

   Special_Conditions_at_Site  Carriageway_Hazards  Urban_or_Rural_Area  \
1                           0                    0                    1   
2                           0                    0                    1   
4                           0                    0                    1   
5                           0                    0                    1   
6                           0                    0                    1   

   Did_Police_Officer_Attend_Scene_of_Accident  LSOA_of_Accident_Location  
1                                            1                  E01002836  
2                                            1                  E01002891  
4                                            1                  E01002876  
5                                            1                  E01002858  
6                                            1                  E01002868  

[5 rows x 33 columns]

### Cambiar tipos de datos necesarios

Necesitamos cambiar los tipos de datos cuando vamos a operar sobre de ellos y los operandos no corresponden al tipo de dato que soporta el operador, por ejemplo, si tenemos los datos de la columna Speed_limit como tipo carácter y queremos sacar la media de los mismos tendremos un problema por lo que primero debemos transformarlo a un tipo númerico para poder operar sobre de ellos.

En nuestro caso nuestro data frame es un su mayoría numérico excepto por las columnas de **Accident_Index** **Date** **Time** **Local_Authority_.Highway.** **LSOA_of_Accident_Location**, corroboramos que los datos de las fechas estén en el formato especifícado:

In [52]:
accidents['Date'].tail()

207385    22/10/2004
207386    28/10/2004
207392    12/11/2004
207398    03/12/2004
207406    22/12/2004
Name: Date, dtype: object

In [53]:
accidents['Time'].head()

1    22:21
2    22:47
4    12:15
5    16:50
6    18:20
Name: Time, dtype: object

El formato coincide de forma correcta, y puesto que hasta este momento no sabemos qué tipo de operaciones vamos a relizar no es prudente cambiar el formato o adaptar los mismos, de caso de ser necesario se haría de la siguiente manera:

In [55]:
# para cambiar el tipo de dato o el formato haciendo uso de expresiones regulares:
# accidents['Column_name'].str.replace('regex', 'char_use_to_replace').astype(int)

# aplicando transformaciones a los datos mediante funciones anónimas, en el ejemplo tomamos cada valor y lo
# elevamos al cuadrado
# accidents['Column_name'].apply(lambda x: x**2)